packages/office-ui-fabric-react/src/components/TextField/TextField.base.tsx

```diff
@@ -149,6 +152,26 @@ export class TextFieldBase extends BaseComponent<ITextFieldProps, ITextFieldStat
     if (newProps.defaultValue !== this.props.defaultValue && newProps.value === undefined) {
       this._setValue(newProps.defaultValue);
     }
+
+    // Text field is changing between single- and multi-line. After the change is complete,
+    // we'll need to reset focus and selection/cursor.
+    if (!!newProps.multiline !== !!this.props.multiline && this.state.isFocused) {
+      this._resetFocusAfterInputTypeChange = true;
+      this._selectionBeforeInputTypeChange = [this.selectionStart, this.selectionEnd];
+    }
+  }
+
+  public componentDidUpdate(): void {
+    if (this._resetFocusAfterInputTypeChange) {
+      // The text field has just changed between single- and multi-line, so we need to reset focus
+      // and selection/cursor.
+      this._resetFocusAfterInputTypeChange = false;
+      this.focus();
+      const [start, end] = this._selectionBeforeInputTypeChange!;
```

Comment:
We often talked about avoiding using `!` as it tends to indicate there is a loose binding that is brittle and can easily break without any type safety. While the side effects here are minor, are there any ways to avoid using it? It seems there is a check to see if start and end are defined anyways, so the check could also encapsulate (or be replaced by) checking `_selectionBeforeInputTypeChange`.

Reply:
+1